File: c/map/ep/ep1.py
import numpy as np
import matplotlib.pyplot as plt
import math
import time
#Função que analisa a possibilidade de alterar a tipagem
def is_float(value):
  try:
    float(value)
    return True
  except:
    return False
#Função que preenche uma matriz nxn
def preencheMat(M):
    n = len(M[0])
    v = [0]*n
    j = 0
    check = False
    num = [0]*n
    c = input('Teste A?(s, n): ')
    if c == 's' or c == 'S':
        text_file = open("testeA.txt", "r")
        lines = text_file.readlines()
        for i in range(n):
            j=0
            num[i]=lines[i]
            for x in num[i]:
                if is_float(x):  
                    M[i][j] = float(x)
                    j+=1
        text_file.close()
    else:
        for i in range(n):
            check = False
            while not check:
                for j in range(n):
                    v[j]=float(input('Digite a posição [%d][%d] da matriz: ' % (i+1, j+1)))
                print(v)
                check = checar()
                if (check):
                        M[i]=v  
#Função que preenche um Vetor
def preencheVet(v):
    n = len(v)
    v1 = [0]*n
    check = False
    while not check:
        check = False
        for i in range(n):
            v1[i]=int(input('Digite o numero da posição %d do vetor v: ' %(i+1)))
        print(v1)
        if(checar()):
            v = v1
            check = True
#Função para checar se o usuário digitou valores corretamente
def checar():
    c = input('Os valores estão corretos?(s, n): ')
    if (c == 's' or c == 'S'):
        return True
    return False
#Função calcula a norma de um vetor 
def norma(v):
    a = 0                   
    for i in range(len(v)):
        a+=v[i]**2
    return math.sqrt(a)
#Função que implementa o método de potencia e retorna o tempo de execução em segundos*/
def potencia(M):
    start = time.time()
    n = len(M[0])
    a=0
    e=0.000001
    nv=[0]*n
    v=[0]*n
    av=[0]*n
    preencheVet(v)
    while(math.fabs(norma(v)-norma(nv))>e):
        for i in range(n):
            for j in range(n):
                a+=M[i][j]*v[j]         #nv[k+1] = Av[k]
            nv[i] = a
            a = 0
        for i in range(n):
            v[i]=nv[i]/norma(nv)        #v[k+1] = nv[k+1]/||nv[k+1]||

    for i in range(n):
        a=0
        for j in range(n):
            a+=v[j]*M[j][i]             #(v[k+1])^T * A
        nv[i]=a
        a=0
        for j in range(n):
            a+=nv[j]*v[j]               #(v[k+1])^T * A * v[k+1]
        av[i]=a                         # a[k+1] = (v[k+1])^T * A * v[k+1]

    print('Vetor calculado: '+ v)
    print('Autovalores: '+ av)

    return time.time() - start
#Função que preenche a Matriz de givens correspondente em torno do eixo Z*/
def preencheabg(a, b, g, M):
    for i in range(len(M[0])):
        for j in range(len(M[0])):
            if(i == j):
                a[i] = M[i][j]
            elif (i == (j-1)):
                g[i] = M[i][j]
            elif (i == (j+1)):
                b[i-1] = M[i][j]
#Função que implementa o Algorítimo QR
#def qr():
#Função que implementa o Algorítimo QR com deslocamentos espectrais
def qrDeslocado(M, a, b, g, nd):
    n = len(M[0])
    e = 0.000001
    mi = 0
    I = np.zeros((n,n), dtype=np.float64)
    for i in range(n):
        for j in range(n):
            if i == j:
                I[i][j] = 1
    k=0
    for m in range (n-1, 0, -1):
        mi = 0
        while math.fabs(b[m-1]**k)>e:
            if k>0:
                d = 0.5 * (M[m-1][m-1] - M[m][m])
                mi = M[m][m] + d - (np.sign(d) + (d==0)) * np.hypot(d, M[m][m-1])
                [Q, R] = QRgen(M - mi*k)
                M = (R @ Q) + mi*I
                V = V @ Q
                k +=1

#Função que implementa Rotação de Givens e retorna o tempo de execução em segundos*/
def givens(M, l, c, x):
    start = time.time()
    n = len(M[0])
    alfa = [0]*n        #Vetor com valores da diagonal principal   
    beta = [0]*(n-1)    #Vetor com valores da diagonal inferior
    gama = [0]*(n-1)    #vetor com valores da diagonal superior
    cos = [0]*(n-1)
    sen = [0]*(n-1)
    nd = [0]*(n-2)      #Vetor com valores da nova diagonal
    ab = 1
    B = np.zeros((n,n), dtype=np.float64)
    B = M
    preencheabg(alfa, beta, gama, M)

    for i in range(n-1):
        ab = math.sqrt(alfa[i]**2 + beta[i]**2)
        cos[i] = alfa[i]/ab
        sen[i] = -beta[i]/ab
    for k in range(n-1):
        n1 = M[l-1][k]
        n2 = M[c-1][k]
        
        B[l-1][k] = (cos[k] * n1) - (sen[k] * n2)
        B[c-1][k] = (sen[k] * n1) + (cos[k] * n2)

        if (l-1) == (k-2):
            nd[k-2] = B[l-1][k]

    print('Q%d * A = \n'%(l), B )
    if l+1<n:
        givens(M, l+1, c+1, x)
        
    
    # if x == 'd':
    #     qrDeslocado(M, B)
    # else:
    #     qr(M, B)

    return time.time() - start

def main():
    n = int(input('Digite o tamanho da matriz: '))
    M=np.zeros((n,n), dtype=np.float64)
    preencheMat(M)
    print('\nMatriz:\n',M,'\n')
    while n>0:
        n = int(input('Escolha qual método irá usar: \n(0) - Sair \n'+
        '(1) - Método da potência\n'+
        '(2) - Rotações de Givens e a fatoração QR de uma matriz tridiagonal: \n'+
        '(3) - Algorítimo QR com deslocamentos espectrais: '))
        if n == 1:
            t1 = potencia(M)
            print("O tempo de execução do Método de potencia foi: %d" % (t1))
        elif n == 2:
            l = int(input('Escolha a linha i da transformação: '))
            c = int(input('Escolha a linha j da transformação: '))
            t1 = givens(M, l, c, 'n')
            print("O tempo de execução do Algorítimo QR foi: %d" % (t1))
        elif n == 3:
            l = int(input('Escolha a linha i da transformação: '))
            c = int(input('Escolha a linha j da transformação: '))
            t1 = givens(M, l, c, 'd')
            print("O tempo de execução do Algorítimo QR deslocado foi: %d" % (t1))
        else:
            print("Operação invalida")

if __name__ == "__main__": main()

File: c/Metodos-Numericos/ep/ep1.c
#include <stdio.h> /*teste no pc*/
#include <math.h>
#include <stdlib.h>
#include <time.h>
#include <stdbool.h>

#define maxcol 4
/*Função que calcula o tamanho de vetores*/
int len(float v[]) {
    printf("%d %d\n", sizeof(v), sizeof(v[0]));
    return sizeof(v)/sizeof(v[0]);
}
/*Função que gera a matriz nxn
void criaMatriz(float *M, int n){ 
    M = (float **) malloc (sizeof(float)*n);
    for (int i = 0; i < n; i++ ) {
        M[i] = (float*) malloc (sizeof(float)*n);
    }
}*/
/*?Função que preenche uma matriz nxn
void preencheMatriz(float M[][maxcol]){
    int n = len(M[0]), j=0;
    char c;
    float *v=NULL;
    v = (float*) malloc(n* sizeof(float));
    for (int i = 0; i < n; i++){
        while(j<n){
            printf("\nDigite a posicao [%d][%d] da matriz: ", i+1, j+1);
            scanf("%d", &v[j]);
            imprimeVet(v);
            printf("A linha esta correta?(s, n): ");
            scanf("%c", &c);
            if (c == "s" || c == "S"){
                M[i] = v;
                j++;
            }
        }
    }
}*/
/*Função que preenche um vetor*/
void preencheVetor(float v[]){
    for (int i=0; i < len(v); i++){
        printf("Digite o numero da posição %d do vetor v: ", i+1);
        scanf("%f", &v[i]);
    }
}
/*Função que Calcula a norma de um vetor*/
float norma(float v[]){
    int a=0;
    for (int i=0; i<len(v); i++){
        a+=pow(v[i], 2);
    }
    return sqrt(a);
}
/*Função que imprime um vetor no terminal*/
void imprimeVet (float v[]){
    for (int i=0; i < len(v); i++){
      printf("%f ", v[i]);
    }
    printf("\n");    
}
/*Função que imprime uma matriz no terminal*/
void imprimeMat (float M[][maxcol]){
    int n = len (M[0]);
    for(int i=0; i<n; i++){
        imprimeVet(M[i]);
        printf("\n");
    }
}
/*Função que implementa o metodo de potencia e retorna o tempo de execução em segundos*/
int potencia(float M[][maxcol]){ 
    time_t start = time(NULL);
    int n = len(M[0]);
    /* av = vetor com os autovalores nv = vetor calculado, v = vetor dado, a = auxiliar, e = erro*/
    float a=0, e=0.000001, *nv = NULL,  *v = NULL, *av=NULL; 
    nv = (float *) malloc(n * sizeof(float));
    v = (float *) malloc(n * sizeof(float));
    av = (float *) malloc(n * sizeof(float));

    preencheVetor(v);
    /*Laço a ser executado enquanto o vetor calculado nao se aproxima do vetor dado por e*/
    while(fabs(norma(v)-norma(nv))>e){
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                a+=M[i][j]*v[j];
            }                       /*nv[k+1] = Av[k]*/
            nv[i]=a;
            a=0;
        }
        for(int i=0; i<n; i++){
            v[i]=nv[i]/norma(nv);   /*v[k+1] = nv[k+1]/||nv[k+1]||*/
        }
    }
    
    for(int i=0; i<n; i++){
        a=0;
        for(int j=0; j<n; j++){
            a+=v[j]*M[j][i];        /*(v[k+1])^T * A*/
        }
        nv[i]=a;
        a=0;
        for(int j=0; j<n; j++){
            a+=nv[j]*v[j];          /*(v[k+1])^T * A * v[k+1]*/
        }
        av[i]=a;                    /*a[k+1] = (v[k+1])^T * A * v[k+1]*/
    }

    imprimeVet(v);
    printf("vetor calculado\n");

    imprimeVet(av);
    printf("Autovalores\n");

    free(v);
    free(nv);
    return time(NULL) - start;
}
/*Função que preenche a Matriz de givens correspondente em torno do eixo Z*/
void preencheabg(float a[], float b[], float g[], float M[][maxcol]){
    int n = len(M[0]);
    for (int i=0; i<n; i++){
        for (int j=0; j<n; j++){
            if(i == j)
                a[i]=M[i][j];
            else if (i == j-1)
                g[i] == M[i][j];
            else if (i == j+1)
                b[i-1]==M[i][j];
        }
    }
}
/*Função que implementa o Algoritimo QR*/
void qr(float M[][maxcol], float B[][maxcol]){
    int n = len(M[0]);
    float *av;/*vetor de autovalores*/
    av = (float *) malloc(n * sizeof(float));


}
/*Função que implementa o Algoritimo QR com deslocamentos espectrais*/
void qrDeslocado(float M[][maxcol], float B[][maxcol]){
    int n = len(M[0]);
    float *av;/*vetor de autovalores*/
    av = (float *) malloc(n * sizeof(float));


}
/*Função que implementa Rotação de Givens e retorna o tempo de execução em segundos*/
int givens(float M[][maxcol], char x){
    time_t start = time(NULL);
    int n = len(M[0]), k=0, i=0, j=0;
    float *alfa=NULL, *beta=NULL, *gama=NULL, *cos=NULL, *sen=NULL, ab, **B;
    alfa = (float *) malloc(n * sizeof(float));
    beta = (float *) malloc((n-1) * sizeof(float));
    gama = (float *) malloc((n-1) * sizeof(float));
    cos = (float *) malloc(n * sizeof(float));
    sen = (float *) malloc(n * sizeof(float));

    B = (float **) malloc (sizeof(float)*n);
    for (int i = 0; i < n; i++ ) {
        B[i] = (float*) malloc (sizeof(float)*n);
    }
    /*criaMatriz(B, n);*/
    preencheabg(alfa, beta, gama, M);

    for (int i=0; i<n; i++){
        ab = sqrt((pow(alfa[i], 2)+pow(beta[i], 2)));
        cos[i] = alfa[i] / ab;
        sen[i] = -beta[i] / ab;
    }

    for (k=0; k<n; k++){
        for (i=0; i<n; i++){
            B[i][k] = cos[k] * M[i][k] - sen[k] * M[i][k];
            for (j; j<n; j++){
                B[j][k] = sen[k] * M[i][k] + cos[k] * M[j][k];
            }
        }
    }

    imprimeMat(M);


    if (x == 'd')
        qrDeslocado(M, B);
    else
        qr(M, B);

    return time(NULL) - start;
}

int main(int argc, char const *argv[]) {
    int n, t1;
    printf("Digite o tamanho da matriz: ");
    scanf("%d", &n);
    float **M;

    M = (float **) malloc (sizeof(float)*n);
    for (int i = 0; i < n; i++ ) {
        M[i] = (float*) malloc (sizeof(float)*n);
    }
    bool check = false;
    char c;
    float *v=NULL;
    v = (float*) malloc(n*sizeof(float));
    printf("%d %d \n", sizeof(v), sizeof(v[0]));
    for (int i = 0; i < n; i++){
        while(check == false){
            for(int j = 0; j <n; j++){
                printf("\nDigite a posicao [%d][%d] da matriz: ", i+1, j+1);
                scanf("%f", &v[j]);
            }
            imprimeVet(v);
            printf("A linha esta correta?(s, n): ");
            scanf(" %c", &c);
            if (c == 's' || c == 'S'){
                M[i] = v;
                check = true;
            }
        }
    }

/*    criaMatriz(M, n);
    preencheMatriz(M);*/

    printf("Escolha qual metodo irá usar: \n(0) - Sair \n(1) - Metodo da potência");
    printf("\n(2) - Rotações de Givens e a fatoração QR de uma matriz tridiagona: ,");
    printf("\n(3) - Algoritimo QR com deslocamentos espectrais: ");
    scanf("%d", &n);

    while (n>0){
        switch(n){
            case 1:
                t1 = potencia(M);
                printf("O tempo de execução do Metodo de potencia foi: %d", t1);

            case 2:
                t1 = givens(M, 'n');
                printf("O tempo de execução do Algoritimo QR foi: %d", t1);  

            case 3:
                t1 = givens(M, 'd');         
                printf("O tempo de execução do Algoritimo QR deslocado foi: %d", t1);  

        }
    printf("Escolha qual metodo irá usar: \n(0) - Sair \n(1) - Metodo da potência \n(2) - Rotações de Givens e a fatoração QR de uma matriz tridiagona: ");
    scanf("%d", n);
    }

    
}    
